daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala

```diff
@@ -459,66 +454,75 @@ class VariableMap private(vTable: Map[GlobalQName, ArrayBuffer[VariableInstance]
+              bindingQName,
+              candidates.keys.map(_.toString).mkString(", "))
+            throw new ExternalVariableException(msg)
           }
         }
       } else {
-        bindingQName
+        vTable.get(bindingQName.toGlobalQName)
       }
 
-    val variableInstances = vTable.get(varQName.toGlobalQName)
-    if (!variableInstances.isDefined)
-      referringContext.schemaDefinitionError("unknown variable %s", varQName)
-    else {
-      val variable = variableInstances.get.last
-      variable.state match {
-        case VariableDefined if (!variable.rd.external) => {
-          referringContext.SDE("Cannot set variable %s externally. State was: %s. Existing value: %s.",
-            variable.rd.globalQName, VariableDefined, variable.value)
-        }
-
-        case VariableUndefined if (!variable.rd.external) => {
-          referringContext.SDE("Cannot set variable %s externally. State was: %s.", variable.rd.globalQName, VariableUndefined)
+    optVariableInstances match {
+      case None => throw new ExternalVariableException("Variable definition not found: " + bindingQName)
+      case Some(variableInstances) => {
+        // This array of VariableInstances comes from the VariableMap that is
+        // part of the DataProcessor before being copied to pstate/ustate when
+        // a parse/unparse is started. So this array should contain only a
+        // single instance
+        Assert.invariant(variableInstances.size == 1)
+        val variable = variableInstances(0)
+        if (!variable.rd.external) {
+          throw new ExternalVariableException("Variable cannot be set externally: " + variable.rd.globalQName)
         }
-
-        case VariableSet => {
-          referringContext.SDE("Cannot externally set variable %s twice. State was: %s. Existing value: %s",
-            variable.rd.globalQName, VariableSet, variable.value)
-        }
-
-        case VariableRead => {
-          referringContext.SDE("Cannot externally set variable %s after reading the default value. State was: %s. Existing value: %s",
-            variable.rd.globalQName, VariableSet, variable.value)
-        }
-
-        case _ => {
-          val value = VariableUtils.convert(newValue.getAnyRef.toString, variable.rd, referringContext)
-          variable.setValue(value)
-          variable.setState(VariableDefined)
+        variable.state match {
+          case VariableDefined | VariableUndefined => {
+            val value =
+              try {
+                variable.rd.primType.fromXMLString(newValue)
+              } catch {
+                case e: InvalidPrimitiveDataException => {
+                  val msg = "Value for variable %s is not a valid %s: %s".format(
+                    variable.rd.globalQName,
+                    variable.rd.primType.globalQName,
+                    newValue)
+                  throw new ExternalVariableException(msg)
```

Comment:
Converting one exception to another like this always suggests a style to me which is to just chain the exceptions. I.e., make a constructor for ExternalVariableException which takes a cause which is another exception. 

There's nothing wrong with this code as you have it. I would have made ExternalVariableException a base class and created 3 or 4 more specific subclasses for each of the cases where you are throwing them. Then the string assembly all ends up in the getMessage of the specific exception classes. The constructors of the specific exception classes would just be passed the relevant objects needed to render a good message. 

This is a style that lends itself to internationalization better than if the main code assembles strings of messages.

Reply:
I started with the approach of having multiple exceptions, but felt it was a bit tedious because the way are japi/sapi would also need to have all those exceptions and have converters from the "internal" exceptions to the "public" API exceptions. Which is also why I just went with not chaining the exception here--I didn't want to have to chain it in the public API. Maybe it's worth it though?

Reply:
Yeah, the redundant layer thing.

We have to rethink the JAPI/SAPI layering, and maybe be just using the API objects internally rather than building this redundant layer around them all. For any exceptions that have to show up on SAPI, why wouldn't we just use those in the runtime1 instead of having runtime1 define them, and having the API have to have encapsulators?

One possibililty is that Daffodil uses the set of 3 or 4 derived exceptions, and that hides the string hacking of messages, and then JAPI/SAPI just have a single ExternalVariableException that encapsulates the internal one as a delegate and calls its getMessage().

Reply:
Yeah, would be nice to solve this JAPI/SAPI separation. It's pretty terrible.

One reason not to return the internal things is so that we can change them if we want without affecting backwards compatibility. I'm not sure that's a great reason.

Another reason for the separation is that we don't publish Java/Scala docs for internal API. So there would be no documentation about these new exceptions.

I do like having a concept of a stable public API being everything in org.apache.daffodil.japi.*, which makes it easier to figure out what's public and what's internal, and thus stable vs potential not stable.

> One possibililty is that Daffodil uses the set of 3 or 4 derived exceptions, and that hides the string hacking of messages, and then JAPI/SAPI just have a single ExternalVariableException that encapsulates the internal one as a delegate and calls its getMessage().

Feels like this might just add extra maintenance burden. I think the only thing that ever needs to catch these exceptions is the API shim layer, so if it's going to just convert 4 different exceptions to one, do we gain much from having 4 separate exceptions. I guess it's more self documenting, and the error messages are in the exception with a few parameters passed in, rather than passing in a fully formed string?

Reply:
I've pushed an update that I think addresses all comments except this one? Any thoughts one what route you prefer? I don't feel strongly one way or the other.

Reply:
I suggest merge this as is, but subsequently put some thought into the JAPI/SAPI rethink and whether this could be modularized in some better way. If we didn't do this "the right way" due to JAPI/SAPI maintenance overheads, then we really do need to rethink those. 

I am expecting we need to rethink the Daffodil APIs anyway given the whole lack-of modularization of back-ends.